app/instrumentation-tests/src/org/commcare/androidTests/LanguagesTest.kt

```diff
@@ -0,0 +1,135 @@
+package org.commcare.androidTests
+
+import androidx.test.espresso.Espresso
+import androidx.test.espresso.Espresso.closeSoftKeyboard
+import androidx.test.espresso.Espresso.onView
+import androidx.test.espresso.action.ViewActions.click
+import androidx.test.espresso.assertion.ViewAssertions.doesNotExist
+import androidx.test.espresso.assertion.ViewAssertions.matches
+import androidx.test.espresso.matcher.ViewMatchers.*
+import androidx.test.ext.junit.runners.AndroidJUnit4
+import androidx.test.filters.LargeTest
+import org.commcare.CommCareApplication
+import org.commcare.dalvik.R
+import org.commcare.utils.CustomMatchers
+import org.commcare.utils.InstrumentationUtility
+import org.junit.Before
+import org.junit.Test
+import org.junit.runner.RunWith
+
+@RunWith(AndroidJUnit4::class)
+@LargeTest
+class LanguagesTest : BaseTest() {
+
+    companion object {
+        const val CCZ_NAME = "languages.ccz"
+    }
+
+    @Before
+    fun setup() {
+        if (CommCareApplication.instance().currentApp == null) {
```

Comment:
`BaseTest::installApp` do just the same right ?

Reply:
Actually there is a very subtle difference between these 2. 
`BaseTest::installApp` method will install the app only when the app is not already installed. 
This method here will install the app every time even if it's already installed. I had to use it to start with a fresh state so that the changes I make in one method doesn't affect the other method.

Reply:
what makes this test special that we are using a different mechanism here than other tests ?

Reply:
The tests in espresso can run in any order. Meaning `testLanguageChanges` can run before/after   `testAppUpdate_dontInterfereLanguageChanges`. 

Now, `testLanguageChanges` [assumes](https://github.com/dimagi/commcare-android/pull/2382/files#diff-df29b73d980cb2f4a06e71e55835a33ddaa78c83b372c5a1e06f26719a6ce217R45) that the device language is English to begin with. 
And `testAppUpdate_dontInterfereLanguageChanges` changes the device language to Hindi when it finishes. 

So if, `testAppUpdate_dontInterfereLanguageChanges` runs before `testLanguageChanges` then `testLanguageChanges` would fail because `"Enter a name:"` text would've been changed.

Reply:
Got it. thanks for the explanation. If you are looking to abstract this better you can choose to pass a `force` flag to the `BaseTest::installApp` method that if true clears the app even if it's already installed. That way we won't need to write this logic again and again for any tests that need this similar setup.